### PR TITLE
update .net core 3.0 RTM

### LINKED ISSUE #1001: Ocelot is not working with .NET Core 3
## Expected Behavior 

the following method  extension
```
public static IOcelotBuilder AddOcelot(this IServiceCollection services);
public static IOcelotBuilder AddOcelot(this IServiceCollection services, IConfiguration configuration);
```
add the Ocelot Gateway to the AspNetCore project.

## Actual Behavior 
throws an exception 

```
System.TypeLoadException
  HResult=0x80131522
  Message=Could not load type 'Microsoft.Extensions.DependencyInjection.MvcJsonMvcCoreBuilderExtensions' from assembly 'Microsoft.AspNetCore.Mvc.Formatters.Json, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.
  Source=Ocelot
  StackTrace:
   at Ocelot.DependencyInjection.OcelotBuilder..ctor(IServiceCollection services, IConfiguration configurationRoot)
   at Ocelot.DependencyInjection.ServiceCollectionExtensions.AddOcelot(IServiceCollection services, IConfiguration configuration)
   at OcelotApiGw.Startup.ConfigureServices(IServiceCollection services) in <path to source code file> line 73
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.InvokeCore(Object instance, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass9_0.<Invoke>g__Startup|0(IServiceCollection serviceCollection)
   at Microsoft.AspNetCore.Hosting.StartupLoader.ConfigureServicesDelegateBuilder`1.<>c__DisplayClass15_0.<BuildStartupServicesFilterPipeline>g__RunPipeline|0(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.Invoke(Object instance, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass8_0.<Build>b__0(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.StartupLoader.ConfigureServicesDelegateBuilder`1.<>c__DisplayClass14_0.<ConfigureServices>g__ConfigureServicesWithContainerConfiguration|0(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConventionBasedStartup.ConfigureServices(IServiceCollection services)
```
possible related to [this](https://github.com/aspnet/AspNetCore/issues/7220)

## Steps to Reproduce the Problem

  1. Try to add ocelot with the method extension.
  
## Specifications

  - Version: 13.5.2
  - Platform: Windows, .NET Core SDK 3.0 preview8

### PR DESCRIPTION
**Pull request by geffzhang**
## Fixes #1001 
- #1001 

Fix https://github.com/ThreeMammals/Ocelot/issues/1001 and update .net core 3.0 RTM

### CODE REVIEW

**Comment by geffzhang** [comment]
ETSDK1045: The current .NET SDK does not support targeting .NET Core 3.0.  Either target .NET Core 2.2 or lower, or use a version of the .NET SDK that supports .NET Core 3.0. [C:\projects\ocelot-fcfpb\test\Ocelot.IntegrationTests\Ocelot.IntegrationTests.csproj]
    0 Warning(s)
    14 Error(s)
Time Elapsed 00:00:01.51

**Comment by catcherwong** [comment]
The .NET SDK version of both continuous-integrations is `2.2.x`.

Maybe we should update the configuration of CI as well.

File: .travis.yml [hunk 1]
```diff
@@ -11,9 +11,9 @@ dist: bionic
 osx_image: xcode9.2
```
Comment (catcherwong) [inline .travis.yml]: The value of osx_image should update to `xcode9.4`.

File: src/Ocelot.Administration/Ocelot.Administration.csproj [hunk 2]
```diff
@@ -29,10 +27,13 @@
...
+    <PackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.66">
       <PrivateAssets>all</PrivateAssets>
     </PackageReference>
-    <PackageReference Include="IdentityServer4.AccessTokenValidation" Version="2.7.0" />
-    <PackageReference Include="IdentityServer4" Version="2.4.0" />
+    <PackageReference Include="IdentityServer4.AccessTokenValidation" Version="3.0.0" />
+    <PackageReference Include="IdentityServer4" Version="3.0.1" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Update="Microsoft.SourceLink.GitHub" Version="1.0.0-beta2-19367-01" />
```
Comment (Varorbc) [inline src/Ocelot.Administration/Ocelot.Administration.csproj]: 这个包的更新应该在这里
https://github.com/ThreeMammals/Ocelot/blob/c8a2144a4bcc11d3bc0b90510117a2bb3ea5654f/Directory.Build.props#L14

**Comment by dennismdejong** [comment]
Please retry the AppVeyor build as .Net Core 3.0 is available right now and wasen't at the moment the build was queued

**Comment by Loongle** [comment]
Expect a code merge

**Comment by catcherwong** [comment]
@dennismdejong  The dotnet sdk of Ocelot in AppVeyor is `2.2.301` and we do not have permission to modify the configuration to upgrade to `3.0.100` manually because it does not like Travis that have a configuration file `.travis.yml` in the project.

@thiagoloureiro Could you please take a look on AppVeyor build?

**Comment by Loongle** [comment]
Look forward to the new release

**Comment by florisrobbemont** [comment]
Any update on this? Really excited to start using Ocelot with .NET Core 3.0

**Comment by thiagoloureiro** [comment]
@catcherwong I don't think I have permission on appveyor to change that. I'll check it out asap!
@TomPallister I think this PR is very important, can you follow with us please?

**Comment by thiagoloureiro** [comment]
@geffzhang can you re-run the build? Please check something to this branch. Let's see if appveyor go ok!

File: src/Ocelot.Cache.CacheManager/Ocelot.Cache.CacheManager.csproj [hunk 3]
```diff
@@ -28,11 +26,14 @@
     <ProjectReference Include="..\Ocelot\Ocelot.csproj" />
   </ItemGroup>
   <ItemGroup>
-    <PackageReference Include="StyleCop.Analyzers" Version="1.1.118">
+    <PackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.66">
```
Comment (thiagoloureiro) [inline src/Ocelot.Cache.CacheManager/Ocelot.Cache.CacheManager.csproj]: No final version for this one compatible with .net core 3?

Comment (geffzhang) [reply src/Ocelot.Cache.CacheManager/Ocelot.Cache.CacheManager.csproj]: yes

**Comment by geffzhang** [comment]
@thiagoloureiro  Need to modify appveyor , set to .net core 3.0 build

**Comment by la-digital** [comment]
Any update on this?

**Comment by taha-eg** [comment]
any updates ??

**Comment by Loongle** [comment]
Can we merge directly without the consent of appveyor ?

**Comment by jmezach** [comment]
I think we can probably update the AppVeyor configuration using the appveyor.yml, at least that's what I've seen in other projects.

**Comment by captnjosh** [comment]
I cloned & referenced this solution to a converted from asp.net 2.1 core to asp.net 3.0 core of https://github.com/johnph/simple-transaction. And it solved the #1001 issue I had. Thx much.
Gateway api-dependencies

**Comment by Leandropintogit** [comment]
Hello guys,
It's been 14 days since the opening of pull request. 

What is missing to merge? This is an important update for ocelot.

let me know if I can help you with anything

thanks.

**Comment by jmezach** [comment]
May I suggest adding an appveyor.yml file in the root of the repository with the following content:

```yaml
image: Visual Studio 2019
```

I believe that the appveyor.yml file in the repository takes precedence over whatever is configured on the AppVeyor side. Hopefully that will at least get us a green build here.

**Comment by jondmcelroy** [comment]
@geffzhang Have you seen @jmezach 's suggestion?

**Comment by geffzhang** [comment]
@jondmcelroy The dotnet sdk of Ocelot in AppVeyor is 2.2.301 and we do not have permission to modify the configuration to upgrade to 3.0.100 manually because it does not like Travis that have a configuration file .travis.yml in the project.

**Comment by vip32** [comment]
are you sure the move to netcoreapp3.0 as the ONLY target framework ist the right way? netstandard2.0 is still needed for your existing consumers. multitargeting then is the way to go.
https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-3.0&tabs=visual-studio#migrate-libraries-via-multi-targeting

the ocelot code then probably needs #if/#else directives, which is some considerable work to get going.
 
image

**Comment by vip32** [comment]
i mean these kind of things should be controlled by #if/#else directives (when multitargeting)
image

an example [here](https://github.com/vip32/Naos/blob/931bdfc83a6896176812a272225262362b983022/src/Configuration.App.Web/NaosExtensions.cs#L24)
image

**Comment by jmezach** [comment]
I agree with @vip32 in that we probably shouldn't break existing users of Ocelot who want to remain on older versions of .NET Core or even .NET Framework users (which is why Ocelot targeted .NET Standard 2.0 I believe). Multi-targeting is definitely the way to go here.

**Comment by jondmcelroy** [comment]
I don't think multi-targeting is necessary here. There will still be a pre dotnet core 3 branch in which fixes can be back ported and released. Anyone who wants to use ocelot before dotnet core 3, will just need to stay on the version 13 release of ocelot.

Precedent in this library is just to upgrade: https://github.com/ThreeMammals/Ocelot/compare/5ac6ae6...75f4f4e

**Comment by Amine-Smahi** [comment]
Any update on the issue please, i have to deploy the app for the client as soon as possible

**Comment by vip32** [comment]
@jondmcelroy if there is a maintainable v13 branch going single targetframework for >v13 should be ok. the way forward is netcore 3 anyway. however the effort to do multitargeting isn't that big, some  fallbacks (#if NETSTANDARD2_0)  are needed which are all well documented in the link from MS above. so maybe it is easier to maintain a single branch (with multiple targetframeworks) instead of 2...

not sure what the right decision is here, as many here i just want to push this netcore3 PR forward :)

**Comment by catcherwong** [comment]
I agree with @jondmcelroy 

Keep netcoreapp3.0 only maybe the right decision here!

**Comment by TomPallister** [comment]
I have upgraded the build image in AppVeyor to VS2019 and the build seemed to work apart from...

```
Test Run Successful.
Total tests: 585
     Passed: 585
 Total time: 13.6371 Seconds
Committing...
No results, this could be for a number of reasons. The most common reasons are:
    1) missing PDBs for the assemblies that match the filter please review the
    output file and refer to the Usage guide (Usage.rtf) about filters.
    2) the profiler may not be registered correctly, please refer to the Usage
    guide and the -register switch.
2019-10-11T05:49:40: Arguments
2019-10-11T05:49:40:  -reports:C:/projects/ocelot-fcfpb/artifacts/UnitTests/coverage.xml
2019-10-11T05:49:40:  -targetdir:C:/projects/ocelot-fcfpb/artifacts/UnitTests
2019-10-11T05:49:40: Executable: C:\projects\ocelot-fcfpb\tools\ReportGenerator.4.3.0\tools\netcoreapp3.0\ReportGenerator.Core.dll
2019-10-11T05:49:40: Working directory: C:\projects\ocelot-fcfpb
2019-10-11T05:49:40: Loading report 'C:\projects\ocelot-fcfpb\artifacts\UnitTests\coverage.xml' 1/1 in memory
2019-10-11T05:49:40: Finished parsing C:\projects\ocelot-fcfpb\artifacts\UnitTests\coverage.xml...
2019-10-11T05:49:40: Parsing of 1 files completed.
2019-10-11T05:49:40: Start merge result 1
2019-10-11T05:49:40: Finished merge result 1
2019-10-11T05:49:40: Initializing report builders for report types: Html
2019-10-11T05:49:40: Analyzing 0 classes
2019-10-11T05:49:40:   Writing report file 'C:/projects/ocelot-fcfpb/artifacts/UnitTests\index.htm'
2019-10-11T05:49:40: Report generation took 0.1 seconds
Uploading test coverage to https://coveralls.io/github/TomPallister/Ocelot
Coverage data uploaded to coveralls.
Sequence Coverage: 0
Cake.exe : An error occurred when executing task 'RunUnitTests'.
At C:\projects\ocelot-fcfpb\build.ps1:241 char:1
+ &$CAKE_EXE $cakeArguments
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (An error occurr...'RunUnitTests'.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
Error: 
One or more errors occurred.
	
Code coverage fell below the threshold of 80%. You can find the code coverage report at artifacts/UnitTests
Command executed with exception: 
```

Let me know if you need me to do anything. Sorry I have not been around! You guys are awesome! I can't believe how much Ocelot has grown.

**Comment by jmezach** [comment]
@TomPallister I'm guessing that the issue you're seeing is caused by OpenCover/opencover#894, which seems to indicate that OpenCover doesn't yet support .NET Core 3.0. There doesn't seem to be a timeline for when that support will come.

I guess we could investigate coverlet as an alternative?

**Comment by TomPallister** [comment]
yeh look for an alternative or help opencover I guess!

**Comment by TomPallister** [comment]
Is there anything in .net core for test coverage yet?

**Comment by jmezach** [comment]
I believe coverlet is integrated into dotnet test these days yes.

Op vr 11 okt. 2019 19:39 schreef Tom Pallister <notifications@github.com>:

**Comment by TomPallister** [comment]
Probably use that then :)

**Comment by jmezach** [comment]
I agree. I'll try to give it a shot when I have some time. Hopefully soon.

Op zo 13 okt. 2019 11:08 schreef Tom Pallister <notifications@github.com>:

**Comment by jmezach** [comment]
@TomPallister So I managed to make the necessary changes and created pull request geffzhang/Ocelot#7 for those changes to go into his branch. I did have to make some implementation changes as unfortunately, because of a bug that crashed the test process. See the PR for more details.

@geffzhang I've only tested these changes locally so far, so I'm not entirely sure it will work on AppVeyor. If you could merge the PR I created we can see if it works on AppVeyor as well.

**Comment by geffzhang** [comment]
> Please retry the AppVeyor build as .Net Core 3.0 is available right now and wasen't at the moment the build was queued

Cake.exe : An error occurred when executing task 'RunUnitTests'.
At C:\projects\ocelot-fcfpb\build.ps1:241 char:1
+ &$CAKE_EXE $cakeArguments
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (An error occurr...'RunUnitTests'.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
Error: 
One or more errors occurred.
	
CoverallsNet: Could not locate executable.
Command executed with exception:

**Comment by jmezach** [comment]
@geffzhang I created another pull request geffzhang/Ocelot#8 which I believe will fix the build for AppVeyor. I have no idea yet why the Travis build is failing though. Will have to dig in deeper.

**Comment by jmezach** [comment]
I've created a separate pull request (geffzhang/Ocelot#9) for the Travis CI build. Not entirely sure that that will fix it, but at least its cleaner ;).

**Comment by jmezach** [comment]
Okay, looks like the fix for AppVeyor has worked since we have a green build now. Coverage information has been uploaded, but I'm not sure I can view those results anywhere. Perhaps @TomPallister can take a look at that.

Unfortunately the Travis build is still broken. Still not sure what's going on there. It seems to be complaining about ReportGenerator, but that's working fine for me on my local machine which at least is the same platform, though not necessarily the same version. I'll keep trying, but if anyone has any suggestions, don't hesitate to chime in ;).

**Comment by jmezach** [comment]
So, it turned out it was fairly obvious why the Travis CI build broke, so I created another pull request (geffzhang/Ocelot#10). I'm fairly confident that will fix the build ;).

**Comment by jmezach** [comment]
Cool, looks like that fixed things. Guess we're only waiting for a review from @TomPallister now ;).

**Comment by florisrobbemont** [comment]
Awesome work everyone! Can't wait to give Ocelot a spin in .NET Core 3.0

**Comment by vip32** [comment]
very nice, good job!

**Comment by florisrobbemont** [comment]
FYI: I've tested the package (Ocelot & Ocelot.Polly) in my project and everything seems to work.

**Comment by jmezach** [comment]
I've used the sources from this pull request with my project as well, which uses Consul under the hood, but it does look like there's an issue there. On startup I'm getting the following exception:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Ocelot.Provider.Consul.ConsulFileConfigurationRepository..ctor(IOcelotCache`1 cache, IInternalConfigurationRepository repo, IConsulClientFactory factory, IOcelotLoggerFactory loggerFactory) in /Users/jmezach/Pr
ojects/open-source/Ocelot/src/Ocelot.Provider.Consul/ConsulFileConfigurationRepository.cs:line 37
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResol
verLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitIEnumerable(IEnumerableCallSite enumerableCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at RR.ApiGateway.Program.Main(String[] args) in /Users/jmezach/Projects/DeVries/RR.ApiGateway/ApiGateway/Program.cs:line 146
An exception occured during startup:
System.NullReferenceException: Object reference not set to an instance of an object.
   at Ocelot.Provider.Consul.ConsulFileConfigurationRepository..ctor(IOcelotCache`1 cache, IInternalConfigurationRepository repo, IConsulClientFactory factory, IOcelotLoggerFactory loggerFactory) in /Users/jmezach/Projects/open-source/Ocelot/src/Ocelot.Provider.Consul/ConsulFileConfigurationRepository.cs:line 37
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResol
```

It looks like the order in which dependencies are being resolved has somehow changed. Not entirely sure how or why yet. Will dig in.

**Comment by jmezach** [comment]
So I've looked into the issue I've described above and I think I figured out what's going wrong. With ASP.NET Core 3.0 a switch was made to use the generic host, instead of the older WebHost model. It looks like the generic host starts `IHostedServices` earlier, meaning that the `FileConfigurationPoller` service is created fairly early in the startup of the application. This class uses the `IFileConfigurationRepository` dependency, which in the case of Consul (using `AddConsul()`) is the `ConsulFileConfigurationRepository`. That in turn uses the `IInternalConfigurationRepository` to get the global configuration, including the hostname and token used to connect to Consul.

Now the problem is that the `IInternalConfigurationRepository` doesn't have any configuration stored yet, until the `UseOcelot()` extension method is called on the `IApplicationBuilder`, but that won't happen until later, causing the `ConsulFileConfigurationRepository` to throw a NullReferenceException because the internal configuration isn't yet available.

I think this will require some refactoring to make this work properly. Do we want to fix this as part of this PR? Or do we create a new one?

**Comment by jmezach** [comment]
After digging in some more I think I have a fairly simple solution to this issue, but I'm not entirely sure that it's the correct way to fix the issue. It works fine for my use case though. See my commit https://github.com/jmezach/Ocelot/commit/a0ff3898c6dd62e2f30f890f068fa6d70140b958 to see what I've changed @thiagoloureiro @TomPallister

**Comment by Leandropintogit** [comment]
Hi jmezach

Netcore project has a opened issue about that.

https://github.com/aspnet/AspNetCore/issues/14585

**Comment by jmezach** [comment]
@Leandropintogit Interesting discussion which I think demonstrates the exact behaviour I'm seeing as well. Doesn't look like it's likely to be changed back though, so I think Ocelot will need to react to this change.

As for my changes, I think they actually make sense. With the original implementation it seems that the `ConsulFileConfigurationRepository` is trying to use the `IInternalConfiguration` that it is itself responsible for retrieving from Consul. That basically leads to a bit of a chicken and egg problem. With my changes the `ConsulFileConfigurationRepository` relies on a piece of statically available configuration which isn't very likely to change once Ocelot has started. In fact, that is exactly what the [documentation](https://ocelot.readthedocs.io/en/latest/features/configuration.html#store-configuration-in-consul) suggest you do; set the service discovery provider information in `ocelot.json` and the rest of the configuration will be read from there.

Still interested to hear what @TomPallister and @thiagoloureiro think about my change though ;).

**Comment by Leandropintogit** [comment]
> After digging in some more I think I have a fairly simple solution to this issue, but I'm not entirely sure that it's the correct way to fix the issue. It works fine for my use case though. See my commit [jmezach@a0ff389](https://github.com/jmezach/Ocelot/commit/a0ff3898c6dd62e2f30f890f068fa6d70140b958) to see what I've changed @thiagoloureiro @TomPallister

@jmezach 
Good job. I just changed the order of assignment of the _configurationKey  variable


` var serviceDiscoveryProvider = fileConfiguration.Value.GlobalConfiguration.ServiceDiscoveryProvider;`
           ` _configurationKey = serviceDiscoveryProvider.ConfigurationKey ?? "InternalConfiguration";`

            `var config = new ConsulRegistryConfiguration(serviceDiscoveryProvider.Host,
                serviceDiscoveryProvider.Port, _configurationKey, serviceDiscoveryProvider.Token);`

            `_consul = factory.Get(config);`

**Comment by Hbib-drif** [comment]
Great work guys , is there an update or the issue is still not fixed ?

**Comment by jesnasali** [comment]
Is there an update?

**Comment by jmezach** [comment]
I made another pull request (geffzhang/Ocelot#11) with my proposed fix as described earlier, including the fix in ordering of variable assignment that @Leandropintogit mentioned. Let me know if you want to take it ;).

**Comment by jmezach** [comment]
Whoops, forgot to fix the tests it seems. Will fix that ;).

**Comment by pranavpatil19** [comment]
is there an issue or the issue is still not fixed ?

**Comment by LGinC** [comment]
I need a new nuget package to fix json issue

**Comment by jmezach** [comment]
@geffzhang I made another PR (geffzhang/Ocelot#12) with the fix for the broken tests.

**Comment by geffzhang** [comment]
@jmezach I've merged your pr， the error： [/home/travis/build/ThreeMammals/Ocelot/test/Ocelot.UnitTests/Ocelot.UnitTests.csproj]
Consul/ConsulFileConfigurationRepositoryTests.cs(58,59): error CS1503: Argument 1: cannot convert from 'Ocelot.Cache.IOcelotCache<Ocelot.Configuration.File.FileConfiguration>' to 'Microsoft.Extensions.Options.IOptions<Ocelot.Configuration.File.FileConfiguration>' [/home/travis/build/ThreeMammals/Ocelot/test/Ocelot.UnitTests/Ocelot.UnitTests.csproj]
Consul/ConsulFileConfigurationRepositoryTests.cs(58,74): error CS1503: Argument 2: cannot convert from 'Ocelot.Configuration.Repository.IInternalConfigurationRepository' to 'Ocelot.Cache.IOcelotCache<Ocelot.Configuration.File.FileConfiguration>' [/home/travis/build/ThreeMammals/Ocelot/test/Ocelot.UnitTests/Ocelot.UnitTests.csproj]
Consul/ConsulFileConfigurationRepositoryTests.cs(146,59): error CS1503: Argument 1: cannot convert from 'Ocelot.Cache.IOcelotCache<Ocelot.Configuration.File.FileConfiguration>' to 'Microsoft.Extensions.Options.IOptions<Ocelot.Configuration.File.FileConfiguration>' [/home/travis/build/ThreeMammals/Ocelot/test/Ocelot.UnitTests/Ocelot.UnitTests.csproj]
Consul/ConsulFileConfigurationRepositoryTests.cs(146,74): error CS1503: Argument 2: cannot convert from 'Ocelot.Configuration.Repository.IInternalConfigurationRepository' to 'Ocelot.Cache.IOcelotCache<Ocelot.Configuration.File.FileConfiguration>' [/home/travis/build/ThreeMammals/Ocelot/test/Ocelot.UnitTests/Ocelot.UnitTests.csproj]
    1666 Warning(s)

**Comment by jmezach** [comment]
Turned out there were some acceptance tests in there as well which were failing because of a condition I missed. Thank god for tests ;). I created yet another PR (geffzhang#13) which should fix those tests as well.

@geffzhang Not sure where you got that error from. I'm guessing you saw older build results or something. I've seen that happen before as well.

**Comment by geffzhang** [comment]
@thiagoloureiro Release a new version of the nuget package

**Comment by pranavpatil19** [comment]
is there an issue ? waiting for new stable release

**Comment by thangchung** [comment]
> @thiagoloureiro Release a new version of the nuget package

We're still waiting for this release for a long time :(

**Comment by kooshan** [comment]
Oh release it already. Jeez!

**Comment by jmezach** [comment]
@pranavpatil19 No issues as far as I know, but we're not able to get this merged. I think we need someone from ThreeMammals to merge this, which seems to include @AbolfazlRajabpour, @anktsrkr, @binarymash, @briansantura, @jps, @philproctor, @thiagoloureiro and @TomPallister.

**Comment by mehranzand** [comment]
When can I get package through the nuget?

**Comment by Mo-ha-mmed** [comment]
Come ooooonn

**Comment by Mo-ha-mmed** [comment]
Hello guys.
Is there an alternative to ocelot for .net core 3.0

**Comment by TomPallister** [comment]
I will get a release out this morning before I go to work!

**Comment by saadshams** [comment]
updated nuget package please?

**Comment by jmezach** [comment]
@saadshams New version is already available on NuGet (version 13.8.0) with compatibility  for .NET Core 3.0. I have already successfully upgraded my project to 3.0 which uses Ocelot.

**Pull request merged**